pkg/pullrequest/scm.go

```diff
@@ -0,0 +1,133 @@
+
+	"golang.org/x/oauth2"
+
+	"github.com/jenkins-x/go-scm/scm/driver/github"
+	"github.com/jenkins-x/go-scm/scm/driver/gitlab"
+	"go.uber.org/zap"
+)
+
+func NewSCMHandler(logger *zap.SugaredLogger, raw, provider, token string) (*Handler, error) {
+	u, err := url.Parse(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if provider == "" {
+		p, err := guessProvider(raw)
+		if err != nil {
+			return nil, err
+		}
+		provider = p
+	}
+
+	var handler *Handler
+	switch provider {
+	case "github":
+		handler, err = githubHandlerFromURL(u, token, logger)
+	case "gitlab":
+		handler, err = gitlabHandlerFromURL(u, token, logger)
+	default:
+		return nil, fmt.Errorf("unsupported pr url: %s", raw)
+	}
+	return handler, err
+}
+
+func githubHandlerFromURL(u *url.URL, token string, logger *zap.SugaredLogger) (*Handler, error) {
+	split := strings.Split(u.Path, "/")
+	if len(split) < 5 {
+		return nil, fmt.Errorf("could not determine PR from URL: %v", u)
+	}
+	owner, repo, pr := split[1], split[2], split[4]
+	prNumber, err := strconv.Atoi(pr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing PR number: %s", pr)
+	}
+
+	client := github.NewDefault()
+	ownerRepo := fmt.Sprintf("%s/%s", owner, repo)
+	h := NewHandler(logger, client, ownerRepo, prNumber)
+	if token != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: token},
+		)
+		h.client.Client = oauth2.NewClient(context.Background(), ts)
+	}
+	return h, nil
+}
+
+func gitlabHandlerFromURL(u *url.URL, token string, logger *zap.SugaredLogger) (*Handler, error) {
+	// The project name can be multiple /'s deep, so split on / and work from right to left.
+	split := strings.Split(u.Path, "/")
+
+	// The PR number should be the last element.
+	last := len(split) - 1
+	prNum := split[last]
+	prInt, err := strconv.Atoi(prNum)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse pr as number from %s", u)
+	}
+
+	// Next we sanity check that this is a correct url. The next to last element should be "merge_requests"
+	if split[last-1] != "merge_requests" {
+		return nil, fmt.Errorf("invalid gitlab url: %s", u)
+	}
+
+	// Next, we rejoin everything else into the project field.
+	project := strings.Join(split[1:last-1], "/")
+	client := gitlab.NewDefault()
+	if token != "" {
+		client.Client = &http.Client{
+			Transport: &gitlabClient{
+				token:     token,
+				transport: http.DefaultTransport,
+			},
+		}
+	}
+	return NewHandler(logger, client, project, prInt), nil
+}
+
+// gitlab client wraps a normal http client, adding support for private-token auth.
+type gitlabClient struct {
+	token     string
+	transport http.RoundTripper
+}
+
+func (g *gitlabClient) RoundTrip(r *http.Request) (*http.Response, error) {
+	r.Header.Add("Private-Token", g.token)
+	return g.transport.RoundTrip(r)
+}
+
+func guessProvider(url string) (string, error) {
+	switch {
+	case strings.Contains(url, "github"):
+		return "github", nil
+	case strings.Contains(url, "gitlab"):
+		return "gitlab", nil
+	}
```

Comment:
We should probably switch on the host, not the entire URL path. This will probably be more accurate for the default cases.